Moved GPIO commands in the core

giig1967g opened the pull request:
Hi all,
- Moved the GPIO commands for Internal pins, PCF and MCP in the core code from PLUGIN_WRITE.
- Created several support functions to avoid repetition of the code.
- The old code is still commented in order to simplify the debug and the comparison between the old code and the new changes.
- Fixed some bugs along the way.
- The commands are in GPIO.cpp and the support functions in ESPEasy.ino:
-- gpio
-- togglegpio
-- pulse
-- longpulse
-- longpulse_ms
-- status
-- monitor
-- unmonitor

As the changes are really substantial, I really would like the community to test before releasing.

giig1967g committed:
Moved GPIO commands in the core

Hi all,
moved the GPIO commands for Internal pins, PCF and MCP in the core code from PLUGIN_WRITE.

Created several support functions to avoid repetition of the code.

The old code is still commented in order to simplify the debug and the comparison between the old code and the new changes.

Fixed some bugs along the way.

As the changes are really substantial, I really would like the community to test before releasing.

giig1967g committed:
Moved GPIO commands in the core

Hi all,
moved the GPIO commands for Internal pins, PCF and MCP in the core code from PLUGIN_WRITE.

Created several support functions to avoid repetition of the code.

The old code is still commented in order to simplify the debug and the comparison between the old code and the new changes.

Fixed some bugs along the way.

As the changes are really substantial, I really would like the community to test before releasing.

giig1967g committed:
Moved GPIO commands in the core

Hi all,
moved the GPIO commands for Internal pins, PCF and MCP in the core code from PLUGIN_WRITE.

Created several support functions to avoid repetition of the code.

The old code is still commented in order to simplify the debug and the comparison between the old code and the new changes.

Fixed some bugs along the way.

As the changes are really substantial, I really would like the community to test before releasing.

giig1967g commented:
Looking inside the code I found several things that should be addressed:
- `inputswitchstate`: this command is present inside P001. I suggest to remove it as it has several other options like `taskvalueset `and `gpio`
- [plugin#gpio#pinstate#xx]: today to get the value of a PIN to be used in the rules, you can use [plugin#gpio#pinstate#xx]. Il calls PLUGIN_REQUEST. It's quite time consuming. I suggest to add another syntax [GPIO#xx] to uniform it to the events; this formula will read from the cache (portStatus structure) so it will be faster, while the [plugin#gpio#pinstate#xx] will read from the actual value of the port using `digitalRead()`.
- `GPIO,port,2`: with this command you define a pin to be an input pin. This option makes the code quite complex and TBH it's not very clear why we should use this command. I suggest to create a command called `setInputMode,gpio,port` and `setOutputMode,gpio,port` to simplify.
- add a new command to create an EVENT on demand os a specific port: sendEvent,gpio,port. This will generate an event with the value in the cache of the port.

TD-er commented on src/ESPEasy.ino:
```diff
@@ -723,16 +723,16 @@ void run50TimesPerSecond() {
   }
   {
     START_TIMER;
-//    PluginCall(PLUGIN_UNCONDITIONAL_POLL, 0, dummyString);
```
There was a recent issue about re-instating the PLUGIN_UNCONDITIONAL_POLL.
Is that still possible, or should we have another call for it?

giig1967g replied:
sure can be reenabled. Just need to understand what it should be doing.

TD-er replied:
See: https://github.com/letscontrolit/ESPEasy/issues/2712

giig1967g replied:
can we leave it for a different PR?

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+  return false;
+}
+
+boolean GPIO_Read_Switch_State(byte pinNumber, byte pinMode) {
```
Please use `bool` instead of `boolean`

giig1967g replied:
ok

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+  boolean success = false;
+  byte portvalue = 0;
+  byte unit = (Par1 - 1) / 16;
+  byte port = Par1 - (unit * 16);
```
No check for ranges in port and unit?
And this looks like a div/mod split, right?

giig1967g replied:
It's not my code. It's the original code.
I just moved it outside the plugin.
Checks are performed in the single commands, in order to throw log errors.

uzi18 commented:
@giig1967g this is what we waiting for long time, but we can't break existing configurations here.
There was some ideas to integrate also gpio expanders into core, maybe it is also good time (as next step)?
Maybe we can warn user about unsupported command used and link to migration docs ?

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+  {
+    port = port - 8;
+    IOBankConfigReg++;
+  }
```
Is this just a single step, or should this be a while instead of an if?

giig1967g replied:
It's not my code. It's the original code.
I just moved it outside the plugin.

giig1967g commented:
> @giig1967g this is what we waiting for long time, but we can't break existing configurations here.
> There was some ideas to integrate also gpio expanders into core, maybe it is also good time (as next step)?

I have moved also the commands of the expanders in the core  :)
> Maybe we can warn user about unsupported command used and link to migration docs ?

So far there are no unsupported commands.

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+    break;
+    case PLUGIN_MCP:
+    case PLUGIN_PCF:
+      returnValue=(port>=1 && port<=128);
```
Have PCF and MCP the same port range?

giig1967g replied:
yes. 128 both

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+/****************************************************************************/
```
Why not have this as a .cpp file in src/src/Gpio/ ?  (also the .h file for it)

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+#else // if defined(ESP8266)
+  pinMode(port, INPUT_PULLUP);
+#endif // if defined(ESP8266)
+}
```
Just wondering, do the expanders have internal pull-up resistors?

uzi18 replied:
@giig1967g @TD-er PCF - no, MCP yes

TD-er replied:
This function does not check for valid GPIO numbers.
For example changing GPIO 6 - 11 on an ESP8266 or 6 - 9 on ESP8285 is a good recipe for a crash or maybe even data corruption on the flash.

Maybe check with `checkValidPortRange` first and let's extend that function to perform the port range checks on ESP82xx.

uzi18 commented:
I did not read code jet, so it was just comments on your post here :)

TD-er commented on src/Scheduler.ino:
```diff
@@ -422,105 +416,66 @@ void process_plugin_task_timer(unsigned long id) {
-
-  digitalWrite(pinNumber, pinStateValue);
+  byte GPIOType      = id & 0xFF;
+  byte pinNumber     = (id >> 8);
```
I wonder why the compiler is not sending out a warning when you shift without bit masking.
I know the old code does this, but better make it explicit:
```c++
byte pinNumber = static_cast<byte>((id >> 8) & 0xFF);
```

giig1967g replied:
only pinNumber should change? Or also GPIOType and/or pinStateValue?

TD-er replied:
The other one also. I just added it as an example.

giig1967g replied:
you mean like this?
byte GPIOType = static_cast<byte>((id) & 0xFF);
byte pinNumber = static_cast<byte>((id >> 8) & 0xFF);
byte pinStateValue = static_cast<byte>((id >> 16) & 0xFF);

TD-er replied:
```c++
byte GPIOType = static_cast<byte>((id) & 0xFF);
byte pinNumber = static_cast<byte>((id >> 8) & 0xFF);
byte pinStateValue = static_cast<byte>((id >> 16) & 0xFF);
```

You forgot the `<byte>` for the static_cast

giig1967g replied:
done

uzi18 commented:
for example `PCFGPIO,port,2` was introduced to "set" gpio as known input, because anicent expander like pcf8574 don't understand such thing like input type and when someone send 0 to such input it will never report I/O change , this resolved lots of problems and hidden bugs

TD-er commented on src/_P001_Switch.ino:
```diff
@@ -459,237 +446,263 @@ boolean Plugin_001(byte function, struct EventStruct *event, String& string)
-              event->sensorType = SENSOR_TYPE_SWITCH;
+                byte output_value;
+                currentStatus.output = new_outputState;
+                boolean sendState = new_outputState;
```
type `boolean` is not a bool, so hard to tell what the negate operator will do here.
Can this be a bool?

giig1967g replied:
this is inside PLUGIN_WRITE. I haven't started looking at it yet.

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,392 @@
+
+          return return_command_success();
+        }
+        break;
```
Has this been taken from older code?
I thought I had changed these break-after-return statements a while ago (not entirely sure though)

giig1967g replied:
I have rebased my fork a week ago. So this should be recent code.
But the lines you are pointing to are my code. 
How should I change it?

TD-er replied:
I remember the static code analyzer tools I used were warning me about having a break after a return statement, which is unreachable code and often not as intended.

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,392 @@
+  			case PLUGIN_PCF:
+  				GPIO_PCF_Write(event->Par1, state);
+  				break;
+      }
```
Why not use your new function `GPIO_Write` here?

giig1967g replied:
Will do. It's not fully optimized yet...

giig1967g replied:
done

giig1967g replied:
done

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,392 @@
+  return success;
+}
+
+bool getPluginIDAndPrefixAndType(char selection, byte* pluginID, String* logPrefix, byte* gpioTimerType)
```
Why use pointers here for the parameters?
They are not optional, right? So better use `&` here.

giig1967g replied:
you are right

giig1967g replied:
done

giig1967g commented:
BTW: the new version is 21kB smaller! 
And the new monitor routine (that runs 10 times per second) is 300% faster!
During a Internal GPIO change it takes 0,5ms instead of 1,5ms

Attached is the compiled version to be tested.
[2778_moved_gpio_cmd_normal_ESP8266_4M1M.zip](https://github.com/letscontrolit/ESPEasy/files/3883992/2778_moved_gpio_cmd_normal_ESP8266_4M1M.zip)

giig1967g commented:
> 
> 
> for example `PCFGPIO,port,2` was introduced to "set" gpio as known input, because anicent expander like pcf8574 don't understand such thing like input type and when someone send 0 to such input it will never report I/O change

But this could be achieved with a specific command. Instead of using `pcfgpio,port,2`, my suggestion was to use `setInputMode,pcf,port.`
Same code behind, just a different syntax.

TD-er commented:
> BTW: the new version is 21kB smaller!

Compared to what version?
That's also the amount of space gained recently with the explicit `lib_ignore` of the SDFS and SDFat libraries in PlatformIO.
Would be really great if we would gain another 20 k in size. (minimal build would then be around 590k)

Also speed gain is good to see, probably because it doesn't have to check all tasks for executing a command. (and some other changes you made)

giig1967g commented:
> 
> 
> > BTW: the new version is 21kB smaller!
> 
> Compared to what version?
> That's also the amount of space gained recently with the explicit `lib_ignore` of the SDFS and SDFat libraries in PlatformIO.

Compared with version of november 14th that I build for #2378. It was 859kB. The new one is 838kB.

TD-er commented:
I rebased your PR on the current mega branch and built the minimal build for core 2.6.1
It is just about the same as without your PR, so it looks like your gain is based on the SDFS libraries.

```
RAM:   [=====     ]  50.7% (used 41528 bytes from 81920 bytes)
Flash: [=======   ]  68.9% (used 615564 bytes from 892912 bytes)
Creating BIN file ".pio\build\minimal_core_261_ESP8266_1M_OTA\firmware.bin" using ".pio\build\minimal_core_261_ESP8266_1M_OTA\firmware.elf"
```

TD-er commented:
I made a full [ESPEasy_mega-20191123-4-PR_2778.zip](https://www.dropbox.com/s/0132kav93p5823d/ESPEasy_mega-20191123-4-PR_2778.zip?dl=0) test build for this PR as it is now.

TD-er commented:
Can you rebase this PR to the current HEAD of the mega branch?
I can also do it if you like, but I'm afraid it will give issues with the rights for you to push commits on to.
When I did it last night to my local branch for this PR, I had 3 merge conflicts which were quite easy to fix.

giig1967g commented:
> 
> 
> I rebased your PR on the current mega branch and built the minimal build for core 2.6.1
> It is just about the same as without your PR, so it looks like your gain is based on the SDFS libraries.
> 
> ```
> RAM:   [=====     ]  50.7% (used 41528 bytes from 81920 bytes)
> Flash: [=======   ]  68.9% (used 615564 bytes from 892912 bytes)
> Creating BIN file ".pio\build\minimal_core_261_ESP8266_1M_OTA\firmware.bin" using ".pio\build\minimal_core_261_ESP8266_1M_OTA\firmware.elf"
> ```

The two minimal versions are not directly comparable, as the new minimal version contains all the commands for the expanders, which the previous version didn't.

giig1967g commented:
> 
> 
> > I rebased your PR on the current mega branch and built the minimal build for core 2.6.1
> > It is just about the same as without your PR, so it looks like your gain is based on the SDFS libraries.
> > ```
> > RAM:   [=====     ]  50.7% (used 41528 bytes from 81920 bytes)
> > Flash: [=======   ]  68.9% (used 615564 bytes from 892912 bytes)
> > Creating BIN file ".pio\build\minimal_core_261_ESP8266_1M_OTA\firmware.bin" using ".pio\build\minimal_core_261_ESP8266_1M_OTA\firmware.elf"
> > ```
> 
> The two minimal versions are not directly comparable, as the new minimal version contains all the commands for the expanders, which the previous version didn't.

I am not sure I am able to do it.

giig1967g commented:
> 
> 
> Can you rebase this PR to the current HEAD of the mega branch?
> I can also do it if you like, but I'm afraid it will give issues with the rights for you to push commits on to.
> When I did it last night to my local branch for this PR, I had 3 merge conflicts which were quite easy to fix.

I am not sure I am able to do it

TD-er commented:
What Git tool do you use?
With Gitkraken it is quite easy to do.
Make sure your editor is closed or else you may run into conflicts caused by file locking.
If you're unsure, make a ZIP of your current working dir, just to be sure.
Have all changed files committed in a commit, since Git doesn't allow to checkout other branches while there are unstaged files.

- Make sure you have the remote repository of letscontrolit added to your view
- Checkout `mega` from `letscontrolit` => Will ask to reset "mega" to the new point => Yes
- Drag your own branch (new/gpiotimers) onto your own "mega" (in the local window)
- Select "rebase 'new/gpiotimers' onto 'mega' "

This will start the rebase and it will come up with a number of merge conflicts (3 in total)


If the Gitkraken route is a bit too much, you can also try to use the rebase option in Github itself.
I don't have this option, as it isn't my repository you're working in, but I can resolve the merge conflicts for you.

giig1967g commented:
> 
> 
> What Git tool do you use?

github desktop

Is this procedure correct?
![immagine](https://user-images.githubusercontent.com/21318099/69530184-e8d1b500-0f71-11ea-8e18-55bba1dd948b.png)

TD-er commented:
I don't know GitHub desktop, but based on the screenshot it looks OK to me.
Your branch only has a single commit and you do want to rebase it onto the upstream/mega one.

giig1967g commented:
I should have done it.

TD-er commented:
Looks good :)

giig1967g committed:
fixed suggestions made by TD-er

giig1967g committed:
fixed suggestions made by TD-er

giig1967g committed:
fixed suggestions made by TD-er

uzi18 commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,393 @@
+//*********************************************************
+
+void GPIO_Monitor10xSec()
+{
```
@giig1967g @TD-er 
I understand this routine will be optimized later as we have here place to improve.
Before globalMapPortStatus loop, we should cache (last value, actual value) here all expenders at once to avoid unnecessary communication.
For every PCF it is one register, for MCP 4 registers.
With this idea we can modify these cached registers when needed and use it if change on inputs.
After loop update registers if any changes. 
This will really reduce I2C communication as for every MCP/PCF GPIO bit it is several 8/16/32 times per chip. 
Be aware this routine will try to fetch data every 0,1s. and should be as fast as possible.
What do You think?

Of course it can be done as next step

TD-er replied:
I totally agree, you should only fetch/write the data for all pins at once.
This can also be done for the internal GPIO's, which is just a matter of reading a 32 bit value (well there are at least 3 values for the ESP internals which you have to combine).
So all actions must be done on a single 32 bit member and also using a bitmask for hiding/selecting what bits are of interest.

But first lets make it all be handled in single functions, so then we can later optimize it. First let it be a working solution.

giig1967g replied:
sure. it's true that this function need to be fully optimized.
I have come out with several ideas:
1) as you said, read and write PCF and MCP in blocks of 8 bit
2) limit the monitor function to monitor ONLY the input pins, assuming that the outut pins can change only using a command and therefore the monitor event could be sent inside the command itself.
3) for input pins, use interrupts

giig1967g commented:
why is the build faiilng?

TD-er commented:
> why is the build faiilng?

```
Compiling .pio/build/minimal_core_242_ESP8266_1M_OTA/src/ESPEasy-Globals.cpp.o
In file included from src/ESPEasy-Globals.h:115:0,
                 from src/ESPEasy-Globals.cpp:3:
/home/travis/.platformio/packages/framework-arduinoespressif8266@2.20402.4-puya/cores/esp8266/core_version.h:3:0: warning: "ARDUINO_ESP8266_RELEASE_2_4_2" redefined [enabled by default]
 #define ARDUINO_ESP8266_RELEASE_2_4_2
 ^
<command-line>:0:0: note: this is the location of the previous definition
Compiling .pio/build/minimal_core_242_ESP8266_1M_OTA/src/ESPEasy.ino.cpp.o
Compiling .pio/build/minimal_core_242_ESP8266_1M_OTA/src/ESPEasyWiFiEvent.cpp.o
In file included from src/ESPEasy-Globals.h:115:0,
                 from /home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:87:
/home/travis/.platformio/packages/framework-arduinoespressif8266@2.20402.4-puya/cores/esp8266/core_version.h:3:0: warning: "ARDUINO_ESP8266_RELEASE_2_4_2" redefined [enabled by default]
 #define ARDUINO_ESP8266_RELEASE_2_4_2
 ^
<command-line>:0:0: note: this is the location of the previous definition
Compiling .pio/build/minimal_core_242_ESP8266_1M_OTA/src/ESPEasy_common.cpp.o
In file included from /home/travis/build/letscontrolit/ESPEasy/src/ESPeasyGPIO.ino:4:0:
src/EspEasyGPIO.h:13:54: error: new declaration 'bool GPIO_Read_Switch_State(EventStruct*)'
 bool GPIO_Read_Switch_State(struct EventStruct *event);
                                                      ^
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:333:9: error: ambiguates old declaration 'boolean GPIO_Read_Switch_State(EventStruct*)'
   NPluginInit();
         ^
In file included from /home/travis/build/letscontrolit/ESPEasy/src/ESPeasyGPIO.ino:4:0:
src/EspEasyGPIO.h:14:57: error: new declaration 'bool GPIO_Read_Switch_State(byte, byte)'
 bool GPIO_Read_Switch_State(byte pinNumber, byte pinMode);
                                                         ^
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:334:9: error: ambiguates old declaration 'boolean GPIO_Read_Switch_State(byte, byte)'
   PluginInit();
         ^
In file included from /home/travis/build/letscontrolit/ESPEasy/src/ESPeasyGPIO.ino:4:0:
src/EspEasyGPIO.h:17:41: error: new declaration 'bool GPIO_MCP_Write(byte, byte)'
 bool GPIO_MCP_Write(byte Par1, byte Par2);
                                         ^
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:336:9: error: ambiguates old declaration 'boolean GPIO_MCP_Write(byte, byte)'
   log += deviceCount + 1;
         ^
In file included from /home/travis/build/letscontrolit/ESPEasy/src/ESPeasyGPIO.ino:4:0:
src/EspEasyGPIO.h:22:41: error: new declaration 'bool GPIO_PCF_Write(byte, byte)'
 bool GPIO_PCF_Write(byte Par1, byte Par2);
                                         ^
/home/travis/build/letscontrolit/ESPEasy/src/ESPEasy.ino:340:9: error: ambiguates old declaration 'boolean GPIO_PCF_Write(byte, byte)'
   log += ')';
         ^
Compiling .pio/build/minimal_core_242_ESP8266_1M_OTA/src/ESPEasy_packed_raw_data.cpp.o
Compiling .pio/build/minimal_core_242_ESP8266_1M_OTA/src/_CPlugin_Helper.cpp.o
*** [.pio/build/minimal_core_242_ESP8266_1M_OTA/src/ESPEasy.ino.cpp.o] Error 1
```
Core 2.4.x stuff.

giig1967g commented:
I noticed that. But how can I fix it?

giig1967g commented:
We still have to decide about the following commands from P001:
-pwm
-servo
-inputswitchstate
-rtttl
-tone

Shall I create these commands in the core, or leave them in P001?
Shall they be valid for expanders too, of just for internal pins (at present thay are only valid for internal pins)?

TD-er commented:
Currently they only operate on internal GPIO pins, so we don't have to decide on that right now.
I think these should be a plugin on their own:
- servo
- rtttl
- tone

I have no idea what this one does:
- inputswitchstate

giig1967g commented:
> Currently they only operate on internal GPIO pins, so we don't have to decide on that right now.
> I think these should be a plugin on their own:
> 
>     * servo
> 
>     * rtttl
> 
>     * tone
> 

so I will not move them.

> I have no idea what this one does:
> 
>     * inputswitchstate
>

This one is an horrible backdoor to change the value of a task from rules, without doing anything else.
The same as taskvalueset, but limited to switch plugin.
In my opinion should be removed.

TD-er commented:
> This one is an horrible backdoor to change the value of a task from rules, without doing anything else.
> The same as taskvalueset, but limited to switch plugin.

Sounds like something used by a controller.

giig1967g commented:
> 
> 
> > This one is an horrible backdoor to change the value of a task from rules, without doing anything else.
> > The same as taskvalueset, but limited to switch plugin.
> 
> Sounds like something used by a controller.

it's an old command, and it makes no sense:
infact it manually forces the output status of an **input** switch disregarding the real state of the switch.
Also, I suspect that the monitor routine will overwrite it's value reading the actual state and generate an event.

See the code:
```
      if (command == F("inputswitchstate")) {
        const uint32_t key = createKey(PLUGIN_ID_001, Settings.TaskDevicePin1[event->Par1]);
        portStatusStruct tempStatus = globalMapPortStatus[key];

        UserVar[event->Par1 * VARS_PER_TASK] = event->Par2;
        tempStatus.output                    = event->Par2;
        tempStatus.command                   = 1;
        savePortStatus(key, tempStatus);
      }
```

uzi18 commented:
inputswitchstate was used to force gpio state, to send this gpio status to controller or update status in controller

have seen lot's of rules with this command

TD-er commented:
@uzi18 "was" as in "no longer" ?

Found this on the [forum](https://www.letscontrolit.com/forum/viewtopic.php?t=2460)
Apparently the numbering of that command was also fishy (at least once)

uzi18 commented:
we were used it in Nettemp controller, sure some people still have this in rules on espeasy

TD-er commented:
So place it in comments for now and if nobody complains it was no longer needed.
I think it is indeed a hack which may no longer be needed (or possible via other means)

giig1967g commented:
The index starts from 0 and not from 1 as in the device page.
Monitor command now superseeded the need to update the status to the controller.

giig1967g commented:
> 
> 
> So place it in comments for now and if nobody complains it was no longer needed.
> I think it is indeed a hack which may no longer be needed (or possible via other means)

ok

giig1967g committed:
deprecating inpuswitchstate

giig1967g committed:
deprecating inpuswitchstate

giig1967g committed:
deprecating inpuswitchstate

uzi18 commented:
It is also kind of solution =)

giig1967g commented:
In the latest version I have added a log entry saying that the command is deprecated.

giig1967g commented:
> 
> 
> I noticed that. But how can I fix it?

how can this be fixed? I don't see the errors in my code.
For example I looked for `#define ARDUINO_ESP8266_RELEASE_2_4_2` but could not find it

giig1967g commented:
Anyone tested this PR?

TD-er commented:
I have one of my nodes running a test build I made of it earlier, but that's not really a proper real-world test.

TD-er commented:
Can you rebase this PR to the HEAD of the mega branch?
There will be a merge conflict in Command.ino, so I think it is best if you do it.

giig1967g commented:
I am on a business trip until end of next week. 
Not sure if I can do it from here.

TD-er commented:
OK, will try to do it later this evening, or tomorrow.

giig1967g commented:
> OK, will try to do it later this evening, or tomorrow.

Hi, I solved the conflicts, but still there are checks that fail. 
Can you help me understand where the problem is?

giig1967g commented:
@TD-er 
any news on this PR?

TD-er commented:
I will do these tomorrow morning (promise)

TD-er commented:
I made a local rebase on top of the current HEAD of the mega branch.
It is now building a test build, which I will test on a few nodes here.

If this does work without any strange issues, I will try to push the branch and make a full test build for people to test. Let's hope that my push will not lock you out.

TD-er committed:
Fix merge issue merging GPIO core handling

TD-er committed:
[GPIO core] Fix bool/boolean function declaration.

TD-er commented:
Last night I made a test build for this PR: [ESPEasy_mega-20191208-25-PR_2778.zip](https://www.dropbox.com/s/5j0yuoum1aqnuc9/ESPEasy_mega-20191208-25-PR_2778.zip?dl=0)

uzi18 commented:
@giig1967g @TD-er do you think we can have alternative for switch plugin to report only "release" event?
it is sometimes far more optimal than "rules" magic to detect short/long press

giig1967g commented:
Hi @uzi18 ,
You are absolutely right.
I already have some ideas in that direction, but we need to wait until this PR is merged and settled before making other changes in those Plugins.

TD-er commented:
Are there any issues with the test build?
If not then I will merge it this evening.
I can then also trigger a new nightly build, or should I wait until I have looked at the static IP => crash issue we now have?

I am almost finished with moving to the new house and my body is refusing almost anything right now.

giig1967g commented:
@TD-er 
Seriously, take your time. 

Personally I would prefer to merge this PR in a build where this PR is the only change in order to clearly identify possible issues and make corrections or rollback.

uzi18 commented:
It is better to merge as is and improve later.
Will take a look on changes made by @giig1967g here.

giig1967g commented:
Hi @uzi18 
the `setGPIOPullupMode `function is called only for internal GPIO pins.
It's not used for neither for MCP neither for PCF.

If the function name is misleading, in the next version I will change the name to `setInternalGPIOPullupMode`

uzi18 commented:
@giig1967g it is just an idea to use internal pullup if available, we already have issue on mcp expander about internal pullup not reconfigured

giig1967g commented:
Moving here the discussion regarding the bugs identified by @mr2c12
1) mqtt controllers and http do not process internal commands (except some of them manually selected)
2) commands only work in lowercase (I cant understand why)

Considering that the SOURCES are:
```
#define VALUE_SOURCE_SYSTEM                 
#define VALUE_SOURCE_SERIAL                 
#define VALUE_SOURCE_HTTP                  
#define VALUE_SOURCE_MQTT                   
#define VALUE_SOURCE_UDP                    
#define VALUE_SOURCE_WEB_FRONTEND           
#define VALUE_SOURCE_RULES    
```

So these SOURCES should execute all the commands:
```
#define VALUE_SOURCE_SYSTEM                 
#define VALUE_SOURCE_SERIAL                 
#define VALUE_SOURCE_WEB_FRONTEND     
#define VALUE_SOURCE_RULES                  
```

While these SOURCES will need to be explicit authorized for every command using a flag.
```
#define VALUE_SOURCE_HTTP                   
#define VALUE_SOURCE_MQTT                   
#define VALUE_SOURCE_UDP                    
```      


The idea could be adding a new parameter to the COMMAND_CASE call.
Current: 
`COMMAND_CASE("pcfgpio", Command_GPIO,2); `
Suggested: 
COMMAND_CASE("pcfgpio", `Command_GPIO,2,VALUE_SOURCE_HTTP|VALUE_SOURCE_MQTT|VALUE_SOURCE_UDP)`

TD-er commented:
I think there will only be 2 sets like these.
Also not entirely sure if you want that check in that macro.
It will then extend the macro to something like `if (valid_source) { ... } else return ...`
And macros will be expanded at compile time, so all extras in that macro will appear in the source code like N times.

On the other hand it does make sense to have a clear list in a single file showing what command does have what permission.

giig1967g commented:
> 
> 
> I think there will only be 2 sets like these.
> Also not entirely sure if you want that check in that macro.
> It will then extend the macro to something like `if (valid_source) { ... } else return ...`
> And macros will be expanded at compile time, so all extras in that macro will appear in the source code like N times.
> 
> On the other hand it does make sense to have a clear list in a single file showing what command does have what permission.

Exactly. But I don't see other solutions, unless we repeat all the commands in another list, with size increase and risk of manual error

giig1967g commented:
Or maybe we could follow a different approach:
we create n arrays (one for every letter of the alphabet) and we store the command name, the function name, the argument number and the flags in this array.
Then when a command is issued we check over this array.

What do you think?

TD-er commented:
Well that idea already is used in the switch statement handling the commands.

But I really don't want to use arrays of structs in the flash.
The main reason for it is that in the (recent) past, it resulted in errors where a build from Linux differs from one made on Windows.

I think you should extend the macro first to get this issue fixed and later I will find a way to make the binary smaller.

giig1967g committed:
Merge branch 'new/gpiotimers' of https://github.com/giig1967g/ESPEasy into new/gpiotimers

TD-er commented on src/_C005.ino:
```diff
@@ -63,6 +63,15 @@ bool CPlugin_005(byte function, struct EventStruct *event, String& string)
+log += cmd;
+log += F(" ;taskIndex=");
+log += event->TaskIndex;
+addLog(LOG_LEVEL_INFO, log);
```
Please use proper indentation and wrap in `if(logLevelActiveFor)...`

giig1967g replied:
work-in-progress

TD-er commented on src/_C005.ino:
```diff
@@ -75,6 +84,15 @@ bool CPlugin_005(byte function, struct EventStruct *event, String& string)
+log += cmd;
+log += F(" ;taskIndex=");
+log += event->TaskIndex;
+addLog(LOG_LEVEL_INFO, log);
```
Same here

giig1967g replied:
work-in-progress

TD-er commented on src/Command.ino:
```diff
@@ -194,8 +194,8 @@ bool executeInternalCommand(const char *cmd, struct EventStruct *event, const ch
           COMMAND_CASE("mqttretainflag", Command_MQTT_Retain,       1); // MQTT.h
-          COMMAND_CASE(       "monitor", Command_GPIO_Monitor,      2); // GPIO.h
 #endif // USES_MQTT
+          COMMAND_CASE(       "monitor", Command_GPIO_Monitor,      2); // GPIO.h
```
Good catch :)

TD-er commented on src/Command.ino:
```diff
@@ -223,7 +223,7 @@ bool executeInternalCommand(const char *cmd, struct EventStruct *event, const ch
           COMMAND_CASE(        "publish", Command_MQTT_Publish,      2); // MQTT.h
 #endif // USES_MQTT
-          COMMAND_CASE(          "pulse", Command_GPIO_Pulse,        2); // GPIO.h
+          COMMAND_CASE(          "pulse", Command_GPIO_Pulse,        3); // GPIO.h
```
Good catch :)

giig1967g commented:
@TD-er 
sorry this version it's my testing version work-in-progress.
I don't know why it has been published.

Please don't consider it.

TD-er commented:
Well it is now part of the PR, so let me know if it is again intended as a PR :)

giig1967g commented:
It's still intended as a PR, but I am still working on it, so the code is not final. 
I wrongly published it, and it's not tested.

Give some more days and I will prepare a final version.

TD-er commented:
Well other development from my side was also stalled way too long due to me moving to another house.
So that may make it easier for you to make something that is mergeable

giig1967g committed:
added selective command execution according to source

Please review the code, as I developed only for C005 so far.

If idea is fine, I will develop for other controllers and http source.

giig1967g commented:
@TD-er 
please tell me if my idea is fine. So far coded for C005.
In case it's fne I will code for other controllers and HTTP source.

giig1967g commented:
@TD-er 
what do you think about my code?

One other thing: I have been looking into C001 (Domoticz controller) but cannot find the function that handles the incoming command. 
In C005 (OpenHAB) the function is `CPLUGIN_PROTOCOL_RECV` but in C001 there is no such function.
Unfortunately, I don't have Domoticz in house to test. Do you have any idea?

TD-er commented:
C001 is Domoticz HTTP, C002 is the MQTT variant.

See the case for `CPLUGIN_PROTOCOL_RECV` in that code.

giig1967g commented:
oops...

giig1967g committed:
added support for C002 (Domoticz MQTT)

giig1967g commented:
Added support for C002.
@TD-er : can you test if it's fine?

TD-er commented on src/_C002.ino:
```diff
@@ -148,14 +148,18 @@ bool CPlugin_002(byte function, struct EventStruct *event, String& string)
+                } else {
+                  ExecuteCommand_plugin(x, VALUE_SOURCE_MQTT, action.c_str());
+                  // trigger rulesprocessing
+                  //SP: WHY WE NEED TO CALL rules FROM HERE?
```
It is sending out an event which may trigger actions in a rules setup.
After all it is a command received from an external source, which can be an event to trigger some action.

giig1967g replied:
Understood. But on C005 (openHAB) there no such call and actions in the rules are triggered anyway.

TD-er replied:
That's for sure something I don't like about the current situation.
Some controllers may require stuff to be done differently, but just the fact that we apparently need special knowledge about some plugins in the controllers is indication of a flaw in the design.
We clearly are missing something in the ESPEasy design and I am not entirely sure what.

For sure we do need some kind of "receiver" as the opposite of a controller we have now, but I am not entirely sure if that's what is missing for the event sending you pointed at here.

I do think that's actually part of the responsibility of the command handling.
So a controller may receive a command (let's assume it is also a task of a controller), but then it should hand it over to a central command handler.

My suggestion is to leave that part out of the scope of your PR as it may open a can of worms and Pandora's box at the same time.
I will add an issue for it.

TD-er replied:
Added a new issue for it: https://github.com/letscontrolit/ESPEasy/issues/2865

giig1967g replied:
ok

giig1967g committed:
added C006, C014 and HTTP support

giig1967g commented:
@TD-er 
I think I have covered all possible cases and controllers.
I suggest that this build should be tested before merging it.

giig1967g committed:
Fixed lowercase bug #2841

giig1967g commented:
Here is test build:
[PR_2778_4M_8266.zip](https://github.com/letscontrolit/ESPEasy/files/4119869/PR_2778_4M_8266.zip)

giig1967g committed:
gpio core commands: fixed couple of bugs

giig1967g committed:
gpio core commands: bug fixing

giig1967g committed:
added asyncevent in ExecuteCommand_all

giig1967g commented:
@TD-er 
Any news on this merge?

TD-er commented:
> @TD-er
> Any news on this merge?

No not really, yet.
The thing is, I have hardly been behind a computer the last few weeks, or at least not for my own free time :(

I really really hope to have some time for ESPEasy the next few days, as I am totally fed up with all the stuff keeping me from working on the things I would like to do (or even paid jobs).

TD-er commented:
I will now have a look at this far too long overdue PR.

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -144,13 +144,13 @@ String Command_GPIO_Pulse(struct EventStruct *event, const char* Line)
   String logPrefix;
   bool caseFound = true;
```
I'm looking through this code and the default value for `casefound` is incorrect.
For example "pr" will be accepted as valid command.

giig1967g replied:
you are right.

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,385 @@
+
+String Command_GPIO_LongPulse_Ms(struct EventStruct *event, const char* Line)
+{
+  String logPrefix = new char;
```
This is a memory leak and should not be used.
Just declare it as `String logPrefix;`

giig1967g replied:
ok

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,385 @@
+
+String Command_GPIO(struct EventStruct *event, const char* Line)
+{
+  String logPrefix = new char;
```
Same here, do not assign `new char` as it is allocating things on the heap and never deletes it.

Also I'm not entirely sure what will happen if that String object never gets a value assigned before it is used, as the return value of `new char` is a pointer to uninitialized data.
So even when it is initialized right after, it could be that the constructor of String considers it a C-style string array and tries to allocate memory for it, which may be as many bytes as it takes to read the first '0' value, or it does access memory it is not allowed to and crashes.

giig1967g replied:
ok

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,385 @@
+      success=false;
+  }
+  return success;
+}
```
This function does not initialize `pluginID` to some value like `INVALID_PLUGIN_ID` and you are not initializing the variables before calling these functions.
So better make sure it is invalid before checking it.

Also you're only checking for the first letter, so what if the command was `pulse` or something else?

Please also start using `pluginID_t` for the pluginID types. (see `src/src/Globals/Plugins.h` )

giig1967g replied:
ok

TD-er commented on src/src/Commands/GPIO.cpp:
```diff
@@ -0,0 +1,385 @@
+      success=false;
+  }
+  return success;
+}
```
This shares quite a bit of code with `getPluginIDAndPrefix`
If one is a convenience function over the other, please let one call the other and discard the info you don't need or the use cases that are not meant to be valid.

One issue I do see is that `getPluginIDAndPrefix` does allow for the 'g' ("GPIO") but `getPluginIDAndPrefixAndType` doesn't.
The other way around for 'l' (Longpulse).
So maybe you want to have a second look at those as this may lead to bugs in the future if it isn't already a bug.

giig1967g replied:
I will review the whole code, but just to explain the idea bihind the check on the first letter is due to two things:
1) I was trying to introduce a way to reduce the comand size and therefore the rules size. 
With these changes the user can write `monitor,g,1` instead of `monitor,gpio,1`. In the future I was planning to introduce short codes for the commands so that the rules file is shorter and therefore faster to parse.

2) I can use the `switch `statement that is fast

TD-er replied:
I totally understand the optimization of the switch by looking at the first character.
I've also done it on the command parsing self.
Maybe you can first make a list of the commands you like to have in this PR?
Maybe we can also just have them defined as a separate command entry?
Or just have a single function to parse the command string and return an enum (a bit the same idea as you did in this function), but now you have two of these functions and that can be reduced to 1 :)

giig1967g replied:
ok. I have removed `getPluginIDAndPrefixAndType `and collapsed to one function.

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,396 @@
+//********************************************************************************
+bool GPIO_Internal_Read(byte pin)
+{
+  return digitalRead(pin)==HIGH;
```
Does not check using `checkValidPortRange`

giig1967g replied:
Can you help me identifying the valid port ranges for each platform?
8266: 0-5 & 12-16
8285: 0-5 & 10-16
ESP32: ??

And I cannot find anywhere `#if defined ESP8285`. 
Is it used?

TD-er replied:
See the file `src/Hardware.ino`
```c++
bool getGpioInfo(int gpio, int& pinnr, bool& input, bool& output, bool& warning) {
```

There is a version for ESP8266/8285 and one for ESP32

giig1967g replied:
ok

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -0,0 +1,396 @@
+void GPIO_Internal_Write(byte pin, byte value)
+{
+  pinMode(pin, OUTPUT);
+  digitalWrite(pin, value);
```
Does not check using `checkValidPortRange`

giig1967g replied:
ok

TD-er reviewed (changes requested):
Took a look at the code for things that stuck out to me as odd or just a bug (`new char` things really should be fixed)

giig1967g committed:
applied changes requested by TD-er

TD-er commented on src/ESPeasyGPIO.ino:
```diff
@@ -322,14 +327,21 @@ void sendMonitorEvent(const char* prefix, byte port, int8_t state)
+      returnValue=((port>=0 && port<=19) || (port>=21 && port<=23) || (port>=25 && port <= 27) || (port>=32 && port<=PIN_D_MAX));
+    #else
+      returnValue=false;
+    #endif //#if defined(ESP8266)
```
Why not call `getGpioInfo` here? There is already all knowledge present for all internal GPIO's

giig1967g replied:
something like this?
```
    case PLUGIN_GPIO:
      int  pinnr = -1;
      bool input, output, warning;
      returnValue=getGpioInfo(port, pinnr, input, output, warning);
    
    break;

```

It just look to me that getGpioInfo is quite a complex function that has many features that are not needed

TD-er replied:
```c++
    case PLUGIN_GPIO:
    {
      int  pinnr = -1;
      bool input, output, warning;
      returnValue=getGpioInfo(port, pinnr, input, output, warning);
    }    
    break;
```

As you will create new variables within the scope of the case, use { }

And yes it does more than strictly needed, but at least the knowledge of what pin is used where is only defined in a single place in the code.
This is the kind of information you for sure only want to have it occur once in the code so if there is ever a bug in there, you can fix it in one place and not run into different situations where the same bug may apply also but only in some very different use cases.

If the function is too complex, then we can split it into 2 parts like "is it valid" and "anything special about the pin?"
But please have it only defined in 1 spot.
As you may know I may be sometimes a bit too OCD about DRY (Don't Repeat Yourself), but for these kind of things where the very specific knowledge is used in a number of places you really must try to keep it defined in only a single spot in the code or else you will loose countless hours debugging and fixing.

giig1967g replied:
done

giig1967g committed:
updated chackValidPortRange function

giig1967g committed:
fixed function

giig1967g commented:
I have performed all the changes requested

TD-er commented:
For now I don't see anything wrong.
I will make a test build for it.

TD-er commented:
Hmm I did start a test build when I mentioned that I would, but forgot to copy it....

Here it is: [ESPEasy_mega-20200204-33-PR_2778.zip](https://www.dropbox.com/s/7hh6askvvrzuk0q/ESPEasy_mega-20200204-33-PR_2778.zip?dl=0)

giig1967g commented:
hi, news about this merge?

TD-er commented:
More like, did anyone test it?

TD-er commented:
@giig1967g As reported here: https://github.com/letscontrolit/ESPEasy/issues/2892#issuecomment-587071587
Can you have a look at the returned JSON vs. "ok" ?

giig1967g commented:
sure, what should be the output? JSON or TEXT?
Because, as far as I understand, currently is both.

TD-er commented:
Current situation of "both" is a bug anyway, since it does yield incorrect JSON.
I would say go for JSON, as long as you don't need to change a lot in how commands are processed.
If that's needed, please let me know as we can then introduce a flag in de commands to output JSON and maybe use that as default for the more elaborate answers like the GPIO commands
In the end we will move to a lot more JSON communications anyway.

giig1967g commented:
ok

giig1967g committed:
fixed JSON/OK return message in HTTP source

Restored previous behaviour:

Following commands will return OK:
event, asyncevent,taskrun,taskvalueset, taskvaluetoggle, let,logportstatus,jsonportstatus and rules

Al other commands will return JSON or whatever is defined in its command logic.

giig1967g commented:
hi @TD-er 
fixed the #2892 request.

There are some conflicts to resolve, but none of them are related to my code. So in order to avoid to mess up the changed you might have made, it's better if you look at the conflicts.

thanks

giig1967g commented:
hi @TD-er 
news on this PR?

TD-er commented:
I have been quite busy the last few days to get the build size down as you may have seen.
Today I also moved a lot of code, which resulted in faster and smaller code.

Tomorrow I will thoroughly test your code and probably merge it (unless I find something that's really not good and not easy to fix myself.

TD-er commented:
I have the last mega branch code running on one Sonoff node and the code of this PR on another one.
Just to be complete, I use the Sonoff POW 4M1M build and have it using the "Factory default" setting for them, including the rules + switch/relay setup generated by this factory reset.

On the current mega branch code, the button + relay act as you would expect, each press on the button toggles the relay state.

On the code in this PR, it switches "on", but never off again.
The state of the switch (button) does change in the devices page, but the relay doesn't toggle.

I noticed the output of the switch state is also different, so maybe that has something to do with it?

![image](https://user-images.githubusercontent.com/3751318/75769955-5185b600-5d47-11ea-8343-18d0c0cf61ba.png)

It should be an integer value.

TD-er commented:
Hmm, I removed the switch tasks and added them again which makes the value an integer again, so we need to add some "cleanup" on the task to check for valid settings. Apparently it is still using an old setting.

But still the rules do not toggle.
The Relay switched off when I updated the switch task settings and after pressing the button, it switched on again, so the hardware is working.

giig1967g commented:
Hi,
I ma not sure I understood.
What is "rules + switch/relay setup"?
When has it been introduced?

TD-er commented:
> Hi,
> I ma not sure I understood.
> What is "rules + switch/relay setup"?
> When has it been introduced?

Over a year ago.
If you enter the Factory Reset page, you will be given a selector to preselect some config:
![image](https://user-images.githubusercontent.com/3751318/75777187-ccee6400-5d55-11ea-8454-10bfe1261b54.png)

That will setup 2 switch tasks + generate the  rules for it.
So it was for me a very simple test to see if it was still working and it should be relatively easy to reproduce.

giig1967g commented:
hi,
I actually don't have any SONOFF device, unfortunately.
Will check the code.

giig1967g commented:
@TD-er 
I am having a bit of an hard time to find where the rules.txt are stored in the project.
Can you give me a hint?

TD-er commented:
The rules for this are generated, so I will give you mine and you can replace the GPIO number at will.

```
on Button1#state do
  if [Relay1#state]=0
    gpio,12,1
  else
    gpio,12,0
  endif
endon
```

Button is at GPIO-0 and relay at GPIO-12.

giig1967g commented:
in your environment, what is the SONOFF that is working?
The "POW R1" or the "POW R2 250"?

TD-er commented:
The "R1" is running this pull request code and the other one is running the current HEAD of the mega branch.
So the one showing "0.00" as switch value in the screenshot is not working.
I now have removed the switch tasks and added them again and it is still not working.

giig1967g commented:
the reason why I am asking is because I noticed that one unit has value name = Switch and the other has value name = state.
In your rule you have "State" .

TD-er commented:
Event matching should be done case insensitive, but to be sure I just tested it and it still does not work.

giig1967g commented:
no, sorry. I wasn't talking about case mismatch.
I was talking about value name that are different in your two units.

See screenshot.
![image](https://user-images.githubusercontent.com/21318099/75791366-c15a6780-5d6c-11ea-907f-1a71599a96e6.png)

TD-er commented:
That must have been a difference in when the config was generated.
The other one (R2 250) has these rules:
```
on Button1#switch do
  if [Button1#switch]=1
    gpio,12,1
  else
    gpio,12,0
  endif
endon
```

And like I said, the initial state change does trigger the relay.
So if the relay is "off", and I press the button, then it will switch "on" but never back.

giig1967g commented:
ok. understood.
Last question before I dig into the code: 
- assuming that Button1#state is on (=0) and relay1 is on (=0) when you press the button1 to switch OFF the relay, the relay does not switch off. But does the Relay1#state changes in the Device Page or it remains always = 0?

TD-er commented:
Relay state does not change, but button state does.
I have the button set to Push Button, active low.

giig1967g commented:
ok. I can repruduce the issue.
The problem seems related to an INPUT SWITCH device that is used as an OUTPUT device.
So the [Relay1#State] value is incorrect.

Will have a look now.

giig1967g commented:
@TD-er :
I am almost there. Found the bug.
Tomorrow will be able to submit the change.

TD-er commented:
Great!
Looking forward to it.

giig1967g committed:
fixed SONOFF setup and #2892 bug

giig1967g commented:
Done.

TD-er commented:
Great! Will test it later this evening.

TD-er commented:
Initial tests are positive.
It does seem to work as I expect them to work.

Will do some tests and then I will merge it tomorrow, unless some very strange things come up.

TD-er commented:
Found another breaking change :(

I've tested it also on 2 other boards running various older versions (2 days old and few weeks old) and those work as expected.

What happens is that I configure the switch in Domoticz and send  the "button" switch state to Domoticz.
If I switch it on/off via Domoticz it works as expected.
But as soon as I press the button on the node, it switches and almost immediately it switches back.
On older builds this works just as expected.

workgroupengineering commented:
Hi, it's a great job, but I have a question?
How should it work if there are multiple instances of the I/O extension like PCF8591 ?

I think we should create a specific category of plugins to manage all the I/O extenders that implement all the suggested commands giig1967g.
At initialization, plugin objects record its capabilities in struct like this:

| Device name | Driver | Pin |  R/W | PWM | Pulse | Monitor | Description 
| ---- | ---- | ---- |  ---- | ---- | --- | ---- | ---- | 
| [Core] | Sonoff | 1 | R/W | 0..1023| Yes | Yes | |
| [Core] | Sonoff | 2 | R/W | 0..1023| Yes | Yes | |
|  ... | ... | .... | ... | ... | ... | ... | |
| [Core] | Sonoff | 16 | R/W | 0..1023 | Yes | Yes | |
| Ex1 | PCF8591 | 1 | R/W | 0..1023| Yes | Yes |
|  ... | ... | .... | ... | ... | ... | ... | |
| Ex1 | PCF8591 | 4 | R/W | 0..1023| Yes | Yes |
| Ex2 | PCF8591 | 1 | R/W | 0..1023| Yes | Yes |
|  ... | ... | .... | ... | ... | ... | ... | |
| Ex2 | PCF8591 | 4 | R/W | 0..1023| Yes | Yes |
| Ex3 | PCA9685 | 1 | W | 0..1023| Yes | no |
| Ex3 | PCA9685 | 16 | W | 0..1023| Yes | no |

### Examples of commands

- Pulse Pin 6 of sonoff ([Core]) for 100ms : `` Pulse,6,HIGH,100 ``
- Pulse Pin 2 of second PCF8591 for 100ms: `` Ex2.Pulse,2,HIGH,100 ``
- Set High Pin 2 of second PCA9685: `` Ex3.gpio,3,HIGH ``
- List of all capabilities : `` gpio ``
- ....

I know the suggested changes would introduce several breaking changes, but initially we could create a compatibility layer and depreciate the old commands.

**P.S.**: The values in the table are for illustrative purposes only, they may be incorrect

giig1967g commented:
Hi
It’s not an issue as you address each expander with its address using jumpers. 
In my setup I have more than one in a single unit.

giig1967g commented:
Hi @TD-er 
did you merge this PR?

TD-er commented:
Nope, there was this change in how it was working with Domoticz. Or did I miss an update from you on that?

TD-er commented:
Looks like I also introduced some build issue when trying to solve merge conflicts.
Apparently it is failing on the `Command_logPortStatus` I may have put the #ifdef wrong there when merging.

giig1967g commented:
> 
> 
> Nope, there was this change in how it was working with Domoticz. Or did I miss an update from you on that?

I can't test in Domoticz.
Is just Domoticz that is not working?
Is there anything in the log?

TD-er commented:
> > Nope, there was this change in how it was working with Domoticz. Or did I miss an update from you on that?
> 
> I can't test in Domoticz.
> Is just Domoticz that is not working?
> Is there anything in the log?

Have been away all day today.
I hope to test it tomorrow and check what is causing this.

giig1967g commented:
Hi Gijs,
I need your help in order to fix the bug you found. 
If you help me we can merge this PR and I can move on with other developments.

TD-er commented:
I will try to pinpoint the issue this evening/night.
As you may also have experienced in Italy (probably even more than we do right now), the world is in lockdown, so it is rather uncertain when I can work with my daughter at home for the next three weeks at least and no fallback on grandparents.
So last week I have been taking paid assignments as I don't know how long I will be without them.

giig1967g commented:
yes, a very weird situation.
Take your time. 
As I am at home in this period, I can dedicate more time to EspEasy.

giig1967g commented:
@TD-er 
just a kind reminder

TD-er committed:
Fix incorrect merge.

TD-er committed:
Fix possible uninitialized current state in GPIO monitoring

giig1967g commented:
hi @TD-er 
news on the Domoticz compatibility?
Anything that I can do to help you?

TD-er commented:
@giig1967g I'm terribly sorry, but I haven't had a thorough look at it right now.
Now with the Corona stuff, I have to watch my daughter when Witske is working.
So those days are effectively filled with interruptions leaving me dead tired in the evening.

giig1967g commented:
don't worry. I hope you are fine and safe.
If there is anything I can do to verify the compatibility, please tell me.

TD-er committed:
Fix merge issue due to renamed Command.ino

TD-er committed:
Fix merge issue due to change to EventValueSource::Enum

TD-er committed:
[Commands] Add event value source check for commands

TD-er committed:
Fix merge issue WiFi process disconnect event should not disable webserver

giig1967g committed:
Fixed merge issues

Fixed merge issues and commented 2nd SendStatus inside ExecuteCommand (InternalCommands.cpp)

TD-er commented on src/src/Commands/InternalCommands.cpp:
```diff
@@ -293,6 +300,7 @@ bool executeInternalCommand(const char *cmd, struct EventStruct *event, const ch
     #endif // ifndef BUILD_NO_DIAGNOSTIC_COMMANDS
         COMMAND_CASE_R(   "settings", Command_Settings_Print, 0); // Settings.h
       }
+      COMMAND_CASE_A("status", Command_GPIO_Status,          2); // GPIO.h
```
Maybe rename this command to "GPIOstatus" as "status" is a bit too generic?

giig1967g replied:
ok for me, but it will be a breaking change, as this command was already in the original plugin.

TD-er replied:
Ah ok, then leave it.

giig1967g committed:
fixed "OK" in JSON in a more elegant way

TD-er committed:
Fix merge issues

TD-er committed:
[GPIO] Fix merge issues with recent scheduler changes.

TD-er committed:
Fix merge issues.

giig1967g committed:
fixing merge issues

fixing merge issues

giig1967g committed:
final clean-up

TD-er committed:
Merge branch 'mega' into new/gpiotimers

TD-er committed:
[Domoticz MQTT] Fix Domoticz helper able to send internal commands

Since the GPIO command now is moved to be an internal command, the Domoticz helper commands were no longer processed as those were only tested as a plugin command, not an internal command.

TD-er committed:
[Domoticz] Document typical use case linking button + relay + Domoticz

TD-er commented:
OK, we (Stefano and I) finally found the issue with this PR and Domoticz.
So I plan on merging this PR on Sunday.
Would be nice if it could be tested before the merge.

A test build is building right now.

TD-er commented:
Test build:

- [ESPEasy_ESP32_mega-20201009-60-PR_2778.zip](https://www.dropbox.com/s/nc9zr4v71aypfxx/ESPEasy_ESP32_mega-20201009-60-PR_2778.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20201009-60-PR_2778.zip](https://www.dropbox.com/s/b13p454i5xzxapu/ESPEasy_ESP82xx_mega-20201009-60-PR_2778.zip?dl=0)

giig1967g commented:
tested. seems fine to me

TD-er committed:
Merge branch 'mega' into new/gpiotimers

livello commented:
| ESP_Easy_mega_20201025_test_ESP8266_4M1M_VCC
-- | --

mosquitto_pub -t 'evg_thermo/gpio/12'  -m "0"
3678796 : Info  : GPIO : port#0: set to 0

Completely not working.

giig1967g commented:
Hi @livello 
which controller are you using?

TD-er commented:
Can you try with this [test build](https://www.dropbox.com/s/dncffe3aipss5s1/PR_3334_dashboard_esp_gpio.zip?dl=0) ?
I think it was fixed in PR #3334 which is not yet included in a nightly build.
